sdk/core/core-amqp/test/requestResponse.spec.ts

```diff
@@ -399,4 +409,117 @@ describe("RequestResponseLink", function() {
       );
     }
   });
+
+  describe("sendRequest clears timeout", () => {
+    const _global = getGlobal();
+    const originalClearTimeout = clearTimeout;
+    afterEach(() => {
+      _global.clearTimeout = originalClearTimeout;
+    });
+
+    it("sendRequest clears timeout after error message", async function() {
+      let wasClearTimeoutCalled = false;
+      _global.clearTimeout = (tid) => {
+        wasClearTimeoutCalled = true;
+        return originalClearTimeout(tid);
+      };
+      const connectionStub = stub(new Connection());
+      const rcvr = new EventEmitter();
+      let req: any = {};
+      connectionStub.createSession.resolves({
+        connection: {
+          id: "connection-1"
+        },
+        createSender: () => {
+          return Promise.resolve({
+            send: (request: any) => {
+              req = request;
+            }
+          });
+        },
+        createReceiver: () => {
+          return Promise.resolve(rcvr);
+        }
+      } as any);
+      const sessionStub = await connectionStub.createSession();
+      const senderStub = await sessionStub.createSender();
+      const receiverStub = await sessionStub.createReceiver();
+      const link = new RequestResponseLink(sessionStub as any, senderStub, receiverStub);
+      const request: AmqpMessage = {
+        body: "Hello World!!"
+      };
+      const testFailureMessage = "Test failure";
+      setTimeout(() => {
+        rcvr.emit("message", {
+          message: {
+            correlation_id: req.message_id,
+            application_properties: {
+              statusCode: 400,
+              errorCondition: null,
+              statusDescription: null,
+              "com.microsoft:tracking-id": null
+            },
+            body: "I should throw an error!"
+          }
+        });
+      }, 0);
+      try {
+        await link.sendRequest(request, { timeoutInMs: 120000, requestName: "foo" });
+        throw new Error(testFailureMessage);
+      } catch (err) {
+        assert.notEqual(err.message, testFailureMessage);
+      }
+      assert.isTrue(wasClearTimeoutCalled, "clearTimeout wasn't called.");
+    });
+
+    it("sendRequest clears timeout after successful message", async function() {
+      let wasClearTimeoutCalled = false;
+      _global.clearTimeout = (tid) => {
+        wasClearTimeoutCalled = true;
+        return originalClearTimeout(tid);
+      };
+      const connectionStub = stub(new Connection());
+      const rcvr = new EventEmitter();
+      let req: any = {};
+      connectionStub.createSession.resolves({
+        connection: {
+          id: "connection-1"
+        },
+        createSender: () => {
+          return Promise.resolve({
+            send: (request: any) => {
+              req = request;
+            }
+          });
+        },
+        createReceiver: () => {
+          return Promise.resolve(rcvr);
+        }
+      } as any);
+      const sessionStub = await connectionStub.createSession();
```

Comment:
It's too bad you have to cast it but I do like this. I need to look at sinon more.

Reply:
This also was copied from an existing test. If you remove the `any`, it complains because there are many more fields required than what we're defining (because we don't need as many for our test).